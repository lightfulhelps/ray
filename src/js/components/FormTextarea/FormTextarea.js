// @flow
import * as React from 'react';
import classNames from 'classnames';

type Props = {
  children: string,
  className?: string,
  isInvalid?: boolean,
  isValid?: boolean,
  size?: 'sm' | 'md' | 'lg',
};

const FormTextarea = ({ children, className, isInvalid, isValid, size, ...other }: Props) => {
  const classes = classNames(
    className,
    'form-control',
    size ? `form-control-${size}` : false,
    {
      'is-invalid': isInvalid,
    },
    { 'is-valid': isValid }
  );

  return (
    <textarea {...other} className={classes}>
      {children}
    </textarea>
  );
};

export default FormTextarea;
