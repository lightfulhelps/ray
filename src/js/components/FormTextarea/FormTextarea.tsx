import * as React from 'react';
import classNames from 'classnames';

interface Props {
  children?: string;
  className?: string;
  isInvalid?: boolean;
  isValid?: boolean;
  size?: 'sm' | 'md' | 'lg';
  [key: string]: any;
}

// Size is an attibute of the textArea html element so there was a conflict with our size props.
// Hence the need to omit this atribute
type TextAreaAttribute = Omit<React.HTMLProps<HTMLTextAreaElement>, 'size'>;

const FormTextarea: React.FC<Props & TextAreaAttribute> = ({
  children,
  className,
  isInvalid,
  isValid,
  size,
  ...other
}: Props) => {
  const classes = classNames(
    className,
    'form-control',
    size ? `form-control-${size}` : false,
    {
      'is-invalid': isInvalid,
    },
    { 'is-valid': isValid }
  );

  return (
    <textarea {...other} className={classes}>
      {children}
    </textarea>
  );
};

export default FormTextarea;
