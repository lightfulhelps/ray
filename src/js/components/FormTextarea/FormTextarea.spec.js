import React from 'react';
import { shallow } from 'enzyme';
import merge from 'lodash/merge';
import FormTextarea from './FormTextarea';

const setup = (overrides = {}) => {
  const props = merge({}, overrides);
  const wrapper = shallow(<FormTextarea {...props} />);

  return { wrapper, props };
};

describe('<FormTextarea />', () => {
  it('should render', () => {
    const { wrapper } = setup();

    expect(wrapper).toMatchSnapshot();
  });

  it('should handle children', () => {
    const { wrapper } = setup({ children: 'Children' });

    expect(wrapper.text()).toBe('Children');
  });

  it('should handle className', () => {
    const { wrapper } = setup();

    wrapper.setProps({ className: 'custom' });

    expect(wrapper.hasClass('form-control')).toBe(true);
    expect(wrapper.hasClass('custom')).toBe(true);
  });

  it('should handle isInvalid', () => {
    const { wrapper } = setup();

    expect(wrapper.hasClass('is-invalid')).toBe(false);

    wrapper.setProps({ isInvalid: true });

    expect(wrapper.hasClass('is-invalid')).toBe(true);
  });

  it('should handle isValid', () => {
    const { wrapper } = setup();

    expect(wrapper.hasClass('is-valid')).toBe(false);

    wrapper.setProps({ isValid: true });

    expect(wrapper.hasClass('is-valid')).toBe(true);
  });

  it('should handle size', () => {
    const { wrapper } = setup();

    wrapper.setProps({ size: 'lg' });

    expect(wrapper.hasClass('form-control-lg')).toBe(true);

    wrapper.setProps({ size: 'sm' });

    expect(wrapper.hasClass('form-control-sm')).toBe(true);
  });

  it('should pass through other props', () => {
    const { wrapper } = setup({ tabIndex: 1, id: 'test' });

    expect(wrapper.prop('tabIndex')).toEqual(1);
    expect(wrapper.prop('id')).toEqual('test');
  });
});
