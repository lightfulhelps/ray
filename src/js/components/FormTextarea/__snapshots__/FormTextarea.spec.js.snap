// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<FormTextarea /> should render 1`] = `
<textarea
  className="form-control"
/>
`;
